### PR TITLE
Add Cluster Health metrics

### LINKED ISSUE #15269: Missing documentation for unreachable_nodes and unreachable_health_endpoints metrics
We should add [documentation](https://docs.cilium.io/en/latest/operations/metrics/#metrics-reference) for the `unreachable_nodes` and `unreachable_health_endpoints` metrics.

### PR DESCRIPTION
**Pull request by h3llix**
See commit logs for details.(a screen shot has also been attached
final

Fixes : #15269

### CODE REVIEW

**Comment by h3llix** [comment]
@pchaigno @joestringer Can you please have a look and tell me if the metrics are listed under correct headings , and how the labels are assigned to corresponding metrices ?

**Comment by pchaigno** [comment]
@h3llix You can check the rendered output with `make render-docs`.

File: Documentation/operations/metrics.rst [hunk 1]
```diff
@@ -196,6 +196,7 @@ Name                                         Labels
 ``endpoint_regenerations_total``             ``outcome``                                        Count of all endpoint regenerations that have completed
 ``endpoint_regeneration_time_stats_seconds`` ``scope``                                          Endpoint regeneration time stats
 ``endpoint_state``                           ``state``                                          Count of all endpoints
+`unreachable_health_endpoints``                                                                 Number of health endpoints that cannot be reached
```
Comment (joestringer) [inline Documentation/operations/metrics.rst]: Looks like there's a ` character missing at the start here.

**Review by joestringer — CHANGES_REQUESTED** [review]
In addition to the feedback from Paul (+ below nit), I think that both of these metrics should be in a new category called something like "cluster health" rather than "endpoints" and "nodes". For the target audience of these docs, it's probably more interesting to see the heading "Cluster health" and figure out which metrics to monitor / alert to determine that cluster health is good rather than trying to piece together the intent from separate categories.

File: Documentation/operations/metrics.rst (hunk 1 above)
Comment (h3llix) [reply Documentation/operations/metrics.rst]: @joestringer @pchaigno thanks for the feedback . And yes this looks appealing that they should lie under a new category . I will fix this and update the pr

File: Documentation/operations/metrics.rst [hunk 2]
```diff
@@ -207,6 +207,16 @@ Name                                       Labels
 ``services_events_total``                                                                     Number of services events labeled by action type
 ========================================== ================================================== ========================================================
 
+Cilium health
```
Comment (pchaigno) [inline Documentation/operations/metrics.rst]: I think this should be:
```suggestion
Cluster health
```

**Review by pchaigno — APPROVED** [review]

**Comment by h3llix** [comment]
@pchaigno documentation for metrics `controllers_failing` and `ip_addresses` do not exist . Is this needed in the docs ?

**Comment by pchaigno** [comment]
> @pchaigno documentation for metrics `controllers_failing` and `ip_addresses` do not exist . Is this needed in the docs ?

Nice spotting! :eyes: Yes, please add them.

**Comment by h3llix** [comment]
> > @pchaigno documentation for metrics `controllers_failing` and `ip_addresses` do not exist . Is this needed in the docs ?
> 
> Nice spotting!  Yes, please add them.

Sure i'll add them .
Sorry this might be a silly question . But should i make a new issue and pr . Or is it fine working in this pr ?

**Comment by pchaigno** [comment]
>  Sorry this might be a silly question . But should i make a new issue and pr . Or is it fine working in this pr ?

Not a silly question. I'd add it in a new commit in this PR and would then update the PR's title to better reflect the new changes.

File: Documentation/operations/metrics.rst [hunk 3]
```diff
@@ -207,6 +207,18 @@ Name                                       Labels
...
 
+Cluster health
+~~~~~~~~~~~~~~
+
+========================================== ================================================== ========================================================
+Name                                       Labels                                             Description
+========================================== ================================================== ========================================================
+``unreachable_nodes``                                                                         Number of nodes that cannot be reached
+``unreachable_health_endpoints``                                                              Number of health endpoints that cannot be reached
+``ip_addresses``                                                                              Number of allocated IP addresses
```
Comment (pchaigno) [inline Documentation/operations/metrics.rst]: That metric seems to have a label.

**Review by pchaigno — CHANGES_REQUESTED** [review]

File: Documentation/operations/metrics.rst (hunk 3 above)
Comment (h3llix) [reply Documentation/operations/metrics.rst]: > @pchaigno @joestringer Can you please have a look and tell me if the metrics are listed under correct headings , and how the labels are assigned to corresponding metrices ?

Sorry but i asked this earlier also . How to find the labels of the metrics ?

Comment (pchaigno) [reply Documentation/operations/metrics.rst]:
```
$ git grep "func NewDesc(" vendor/github.com/prometheus/
vendor/github.com/prometheus/client_golang/prometheus/desc.go:func NewDesc(fqName, help string, variableLabels []string, constLabels Labels) *Desc {
```
I'm not sure if `variableLabels` and `constLabels` both need to be documented but checking existing entries should answer that question. Then:
```
$ git grep -C3 ip_addresses pkg/metrics/
pkg/metrics/status.go-                  nil, nil,
pkg/metrics/status.go-          ),
pkg/metrics/status.go-          ipAddressesDesc: prometheus.NewDesc(
pkg/metrics/status.go:                  prometheus.BuildFQName(Namespace, "", "ip_addresses"),
pkg/metrics/status.go-                  "Number of allocated IP addresses",
pkg/metrics/status.go-                  []string{"family"}, nil,
pkg/metrics/status.go-          ),
$ git grep -C3 unreachable_nodes pkg/metrics/
pkg/metrics/status.go-                  []string{"family"}, nil,
pkg/metrics/status.go-          ),
pkg/metrics/status.go-          unreachableNodesDesc: prometheus.NewDesc(
pkg/metrics/status.go:                  prometheus.BuildFQName(Namespace, "", "unreachable_nodes"),
pkg/metrics/status.go-                  "Number of nodes that cannot be reached",
pkg/metrics/status.go-                  nil, nil,
pkg/metrics/status.go-          ),
$ git grep -C3 unreachable_health_endpoints pkg/metrics/
pkg/metrics/status.go-                  nil, nil,
pkg/metrics/status.go-          ),
pkg/metrics/status.go-          unreachableHealthEndpointsDesc: prometheus.NewDesc(
pkg/metrics/status.go:                  prometheus.BuildFQName(Namespace, "", "unreachable_health_endpoints"),
pkg/metrics/status.go-                  "Number of health endpoints that cannot be reached",
pkg/metrics/status.go-                  nil, nil,
pkg/metrics/status.go-          ),
$ git grep -C3 controllers_failing pkg/metrics/
pkg/metrics/status.go-          ciliumClient: ciliumClient,
pkg/metrics/status.go-          healthClient: healthClient,
pkg/metrics/status.go-          controllersFailingDesc: prometheus.NewDesc(
pkg/metrics/status.go:                  prometheus.BuildFQName(Namespace, "", "controllers_failing"),
pkg/metrics/status.go-                  "Number of failing controllers",
pkg/metrics/status.go-                  nil, nil,
pkg/metrics/status.go-          ),
```

File: Documentation/operations/metrics.rst [hunk 4]
```diff
@@ -207,6 +207,19 @@ Name                                       Labels
...
 
+Cluster health
+~~~~~~~~~~~~~~
+
+========================================== ================================================== ========================================================
+Name                                       Labels                                             Description
+========================================== ================================================== ========================================================
+``unreachable_nodes``                                                                         Number of nodes that cannot be reached
+``unreachable_health_endpoints``                                                              Number of health endpoints that cannot be reached
+``ip_addresses``                           ``family``                                         Number of allocated IP addresses
```
Comment (pchaigno) [inline Documentation/operations/metrics.rst]: Maybe the IPAM section would be a better place for this metric? /cc @joestringer

**Review by pchaigno — APPROVED** [review]

File: Documentation/operations/metrics.rst [hunk 5]
```diff
@@ -207,6 +207,19 @@ Name                                       Labels
...
+``controllers_failing``                                                                       Number of failing controllers
+
```
Comment (qmonnet) [inline Documentation/operations/metrics.rst]:
```suggestion
```
No need for this empty line in the table.
Looks good from my side otherwise.

**Review by qmonnet — CHANGES_REQUESTED** [review]

**Review by qmonnet — APPROVED** [review]
Looks good, thank you!

File: Documentation/operations/metrics.rst (hunk 4 above)
Comment (joestringer) [reply Documentation/operations/metrics.rst]: Yes, please put `ip_addresses` under IPAM.

**Review by joestringer — CHANGES_REQUESTED** [review]

**Review by joestringer — APPROVED** [review]

**Pull request merged**